### PR TITLE
Changing type and name of local_used_maps to reflect that it is only one map

### PR DESCRIPTION
**Pull request by jaceyca**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65380

Fixing bugs that arise when running setup.py develop

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @gcramer23

Differential Revision: [D31104844](https://our.internmc.facebook.com/intern/diff/D31104844)

### CODE REVIEW

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 1]
```diff
@@ -643,53 +640,52 @@ void Reducer::autograd_hook(size_t index) {
...
+    TORCH_INTERNAL_ASSERT(
+        local_used_map_tmp.data_ptr() != local_used_map_.data_ptr());
+    local_used_map_tmp.copy_(local_used_map_);
+    local_used_map_dev_.copy_(local_used_map_tmp, true);
+  } else {
+    local_used_map_dev_.copy_(local_used_map_, true);
   }
-  local_used_work_ = process_group_->allreduce(local_used_maps_dev_);
+  std::vector<at::Tensor> temp_local_used_map_dev_vec_ = {local_used_map_dev_};
+  local_used_work_ = process_group_->allreduce(temp_local_used_map_dev_vec_);
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp:688]: curious, does the following work?

```c++
local_used_work_ = process_group_->allreduce({local_used_map_dev_});
```

**Review by mrshenli — APPROVED** [review]
Thanks for further cleaning up the code. 

> Fixing bugs that arise when running setup.py develop

Which bug are you referring to?

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 1 above)
Comment (jaceyca) [reply torch/csrc/distributed/c10d/reducer.cpp:688]: Nope, that causes the `python setup.py develop` command to error out:
```
../torch/csrc/distributed/c10d/reducer.cpp: In member function ‘void c10d::Reducer::all_reduce_local_used_map()’:
../torch/csrc/distributed/c10d/reducer.cpp:688:67: error: no matching function for call to ‘c10d::ProcessGroup::allreduce(at::Tensor&)’
   local_used_work_ = process_group_->allreduce(local_used_map_dev_);
                                                                   ^
In file included from ../torch/csrc/distributed/c10d/reducer.hpp:13,
                 from ../torch/csrc/distributed/c10d/reducer.cpp:1:
../torch/csrc/distributed/c10d/ProcessGroup.hpp:212:50: note: candidate: ‘virtual c10::intrusive_ptr<c10d::ProcessGroup::Work> c10d::ProcessGroup::allreduce(std::vector<at::Tensor>&, const c10d::AllreduceOptions&)’
   virtual c10::intrusive_ptr<ProcessGroup::Work> allreduce(
                                                  ^~~~~~~~~
../torch/csrc/distributed/c10d/ProcessGroup.hpp:212:50: note:   no known conversion for argument 1 from ‘at::Tensor’ to ‘std::vector<at::Tensor>&’
```
This is the bug I was referring to in my commit message, I should've been more clear.

Yanli suggested this fix of pushing `local_used_maps_dev_` into a temporary vector<tensor> and passing it to allreduce()

**Comment by jaceyca** [comment]
@jaceyca has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31104844).

**Pull request closed without merging**